### ISSUE TITLE
Invalid rubocop.yml breaks ale

### ISSUE THREAD

**Issue by kriansa** [w0rp/ale#760]
When using a invalid `rubocop.yml` on your project, Rubocop will output some warning messages, as below:

```
/my/project/.rubocop.yml: Layout/MultilineOperationIndentation has the wrong namespace - should be Style
/my/project/.rubocop.yml: Layout/AlignHash has the wrong namespace - should be Style
{"metadata":{"rubocop_version":"0.48.1","ruby_engine":"ruby","ruby_version":"2.4.1","ruby_patchlevel":"111","ruby_platform":"x86_64-darwin16"},"files":[]}
```

But ale will not parse it right, as it's not a valid JSON anymore, hence it will output some messages like:

```
Error detected while processing function <SNR>168_NeoVimCallback[29]..<SNR>161_HandleExit[40]..ale_linters#ruby#rubocop#Handle:
line    9:
E684: list index out of range: 0
Press ENTER or type command to continue
Error detected while processing function <SNR>168_NeoVimCallback[29]..<SNR>161_HandleExit[40]..ale_linters#ruby#rubocop#Handle:
line    9:
E15: Invalid expression: l:errors['files'][0]['offenses']
```

What should be the right thing to do here?

**Comment by elebow** [comment]
There are a few options that I can see.

1. Suppress the error from the tool. ` 2> /dev/null` appears to work, but I don't know if it's possible on Windows.
1. Discard any non-JSON prefix in the tool's output. Maybe discard any line that does not begin with `{`.
1. Abort parsing gracefully if the output is not valid JSON and add no errors in ALE.
1. Detect the error and notify the user.

Personally, I like choices 1-3, even though they could allow a misconfiguration to go unnoticed by the user.

**Comment by w0rp** [comment]
We should do that same as is done for eslint here. Look for configuration parsing errors first, and report them on line one with a message about using `:ALEDetail` to see the full output.

**Comment by w0rp** [comment]
@kriansa Can you see that output you pasted above in `:ALEInfo` right now? We may or may not have to change the linter to capture both stdout and stderr by changing the option for that.

**Comment by kriansa** [comment]
Sure. Here it is:

```
 Current Filetype: ruby
Available Linters: ['brakeman', 'reek', 'rubocop', 'ruby']
  Enabled Linters: ['brakeman', 'reek', 'rubocop', 'ruby']
 Linter Variables:

let g:ale_ruby_brakeman_options = ''
let g:ale_ruby_reek_show_context = 0
let g:ale_ruby_reek_show_wiki_link = 0
let g:ale_ruby_rubocop_executable = 'rubocop'
let g:ale_ruby_rubocop_options = ''
 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(finished - exit code 0) ['/bin/zsh', '-c', '''rubocop'' --format json --force-exclusion  --stdin ''/project/routes.rb'' < ''/var/folders/hq/4kclj9ds06l060zvml4dmm8m2wk9kf/T/nvimsmWdsm/5/routes-v2.rb''']

<<<OUTPUT STARTS>>>
{"metadata":{"rubocop_version":"0.48.1","ruby_engine":"ruby","ruby_version":"2.4.1","ruby_patchlevel":"111","ruby_platform":"x86_64-darwin16"},"files":[],"summary":{"offense_count":0,"target_file_count":0,"inspected_file_count":0}}
<<<OUTPUT ENDS>>>

(finished - exit code 0) ['/bin/zsh', '-c', 'ruby -w -c -T1 ''/var/folders/hq/4kclj9ds06l060zvml4dmm8m2wk9kf/T/nvimsmWdsm/6/routes.rb''']

<<<NO OUTPUT RETURNED>>>
```

**Comment by w0rp** [comment]
Cheers! Looks like the errors go to stderr, which makes sense.

**Comment by w0rp** [comment]
Now ALE should report problems with rubocop configuration files, like it does for ESLint.

**Comment by kriansa** [comment]
Hello @w0rp,

Thanks for the quick response!

However, it seems that the issue is not really what I suspected. I hadn't gone so far on my debugging when I reported this issue. 😞 

The symptom was that my Neovim was showing some weird messages on random Ruby files, such as:

```
Error detected while processing function <SNR>168_NeoVimCallback[29]..<SNR>161_HandleExit[40]..ale_linters#ruby#rubocop#Handle:
line    9:
E684: list index out of range: 0
Press ENTER or type command to continue
Error detected while processing function <SNR>168_NeoVimCallback[29]..<SNR>161_HandleExit[40]..ale_linters#ruby#rubocop#Handle:
line    9:
E15: Invalid expression: l:errors['files'][0]['offenses']
```

This was not really an issue with `rubocop` being misconfigured - that was a false positive. Actually, ale were working just fine even when those messages are outputed, because they are already being redirected to the stderr. Now after 588595419739e3fec9cad921824e0e07aadb440d it is not.

So I went further debugging it and I found out that the real issue is that sometimes you have ignored files on your `rubocop` settings, and when you run the linter against them, rubocop will output something similar to:

```json
{"metadata":{"rubocop_version":"0.48.1","ruby_engine":"ruby","ruby_version":"2.4.1","ruby_patchlevel":"111","ruby_platform":"x86_64-darwin16"},"files":[],"summary":{"offense_count":0,"target_file_count":0,"inspected_file_count":0}}
```

Notice the empty `files` key there. Currently, ale is assuming that `files` always return at least one value [as you can see here](https://github.com/w0rp/ale/blob/588595419739e3fec9cad921824e0e07aadb440d/ale_linters/ruby/rubocop.vim#L33).

I opened a PR to fix that and to revert 588595419739e3fec9cad921824e0e07aadb440d on #768.

I'm sorry for that :sob:, and thank you for the great work on `ale`!

**Comment by w0rp** [comment]
That other issue has been fixed in #767 now.
